Add used extensions to composer.json

krsriq opened the pull request:
### What is the reason for this PR?

Faker uses a couple of php extensions that are not included in composer.json's require.

- [ ] A new feature
- [ ] Fixed an issue (resolve #ID)

### Author's checklist

- [x] Follow the [Contribution Guide](https://github.com/FakerPHP/Faker/blob/main/.github/CONTRIBUTING.md)
- [ ] New features and changes are [documented](https://github.com/FakerPHP/fakerphp.github.io)

### Summary of changes

This PR adds the required extensions:

- ext-dom is used e.g. in 
https://github.com/FakerPHP/Faker/blob/9f12db8c85c5d10ec1ca02be7097aabd5e1ee67c/src/Faker/Provider/HtmlLorem.php#L50

- ext-iconv is used in 
https://github.com/FakerPHP/Faker/blob/9f12db8c85c5d10ec1ca02be7097aabd5e1ee67c/src/Faker/Provider/ru_RU/Text.php#L11

- ext-json is used in 
https://github.com/FakerPHP/Faker/blob/9f12db8c85c5d10ec1ca02be7097aabd5e1ee67c/src/Faker/Provider/Miscellaneous.php#L321

### Review checklist

- [ ] All checks have passed
- [ ] Changes are approved by maintainer

GrahamCampbell commented:
Kinda too late to do this, since composer will just resolve an older version of this package which it thinks doesn't need these extensions. What would be reasonable though, is for us to add these extensions to the `suggest` and `require-dev` list, instead of `require`.

GrahamCampbell commented on composer.json:
```diff
@@ -14,7 +14,10 @@
-        "php": "^7.1 || ^8.0"
+        "php": "^7.1 || ^8.0",
+        "ext-dom": "*",
+        "ext-iconv": "*",
```
Could we depend on the symfony polyfill?

pimjansen replied:
I wonder where we need them as well tbh?

pimjansen commented:
@GrahamCampbell i actually don't really mind in this case where we now ship a potential broken product.

If we highlight this a bit extra in the release notes i think it is fine.

For the people who want to still bypass they can always ignore the platform requirements while installing

IonBazan commented:
The reason `ext-json` is required by Emoji provider is to support PHP5 which lacks [Unicode Codepoint Escape Syntax](https://wiki.php.net/rfc/unicode_escape) (eg. `"\u{1F600}"`). Since we are already requiring PHP 7.1, we can safely replace it with the new syntax and avoid decoding the json just to show an emoji. Since all files should be encoded in UTF-8, we can actually replace the emoji with actual Unicode characters - just like it is done in https://github.com/jawira/emoji-catalog/blob/master/src/Emoji.php. Please note that https://github.com/spatie/emoji/blob/master/src/Emoji.php uses codepoint escape syntax though.

bramceulemans commented:
I agree with going the `suggests` way. It it not strictly required if you are not using the Providers that use `ext-dom` for example.

```json
{
    "suggest": {
        "ext-dom": "Required for generating random HTML"
    }
}
```

krsriq commented:
I've now moved the extensions to the `composer.json`'s `suggest` section and added a check if `DOMDocument` and `json_decode` are available when they are used.

> The reason ext-json is required by Emoji provider is to support PHP5 which lacks Unicode Codepoint Escape Syntax (eg. "\u{1F600}"). Since we are already requiring PHP 7.1, we can safely replace it with the new syntax and avoid decoding the json just to show an emoji.

I'll do that in a separate PR.

GrahamCampbell commented:
> I'll do that in a separate PR.

Better to do that before this PR, actually.

IonBazan commented on composer.json:
```diff
@@ -34,6 +34,11 @@
     },
+    "suggest": {
+        "ext-dom": "(\\Faker\\Provider\\HtmlLorem) Required for generating random HTML.",
+        "ext-iconv": "(\\Faker\\Provider\\ru_RU\\Text::realText()) Required for generating real Russian text.",
```
Let’s use Symfony Polyfill instead.

krsriq replied:
Couldn't find a Symfony Polyfill for `ext-dom`. So should we keep `ext-dom` in `suggest` and replace `ext-iconv` with `symfony/polyfill-iconv` (but in `required`)?

IonBazan replied:
Yes, I meant only the iconv 😉

pimjansen commented:
@krsriq can we finalize this PR?

localheinz commented on src/Faker/Provider/HtmlLorem.php:
```diff
@@ -47,6 +47,10 @@ public function __construct(Generator $generator)
     public function randomHtml($maxDepth = 4, $maxWidth = 4)
     {
+        if (!class_exists('DOMDocument', false)) {
+            return new \RuntimeException('DOMDocument is required to use randomHtml.');
```
```suggestion
            throw new \RuntimeException('DOMDocument is required to use randomHtml.');
```

Can we also add this to the DocBlock, if we insist on throwing an exception here?

krsriq replied:
Done, thanks!

localheinz commented on src/Faker/Provider/Miscellaneous.php:
```diff
@@ -318,6 +318,10 @@ public static function currencyCode()
     public static function emoji()
     {
+        if (!function_exists('json_decode')) {
+            return new \RuntimeException('The JSON extension is required to generate Emojis.');
```
```suggestion
            throw new \RuntimeException('The JSON extension is required to generate Emojis.');
```

Can we also add this to the DocBlock, if we insist on throwing an exception here?

IonBazan replied:
@krsriq  Since we loosen the BC check rules to allow changing protected properties, it should be now possible to use it without json extension (#124)

krsriq replied:
Yes, we don't need `json_decode` any more.

localheinz commented on src/Faker/Provider/HtmlLorem.php:
```diff
@@ -43,10 +43,16 @@ public function __construct(Generator $generator)
      */
     public function randomHtml($maxDepth = 4, $maxWidth = 4)
     {
+        if (!class_exists('DOMDocument', false)) {
```
```suggestion
        if (!class_exists(\DOMDocument::class, false)) {
```

localheinz commented on src/Faker/Provider/HtmlLorem.php:
```diff
@@ -43,10 +43,16 @@ public function __construct(Generator $generator)
     public function randomHtml($maxDepth = 4, $maxWidth = 4)
     {
+        if (!class_exists('DOMDocument', false)) {
+            throw new \RuntimeException('DOMDocument is required to use randomHtml.');
```
```suggestion
            throw new \RuntimeException('ext-dom is required to use randomHtml.');
```

GrahamCampbell commented on composer.json:
```diff
@@ -34,6 +34,10 @@
     },
+    "suggest": {
+        "ext-dom": "(\\Faker\\Provider\\HtmlLorem) Required for generating random HTML.",
+        "symfony/polyfill-iconv": "(\\Faker\\Provider\\ru_RU\\Text::realText()) Required for generating real Russian text."
```
No point in suggesting a the polyfil. Either require the polyfill, or suggest the extension. I'd vote for the 2nd option, given the English faker stuff doesn't need those functions.

krsriq replied:
Thanks, I changed this now to suggest ext-intl.

GrahamCampbell commented on src/Faker/Provider/HtmlLorem.php:
```diff
@@ -43,10 +43,16 @@ public function __construct(Generator $generator)
      * @param int $maxDepth
      * @param int $maxWidth
      *
+     * @throws \RuntimeException
```
Throws doc doesn't need to document `RuntimeException`s or `Error`s. Code is assumed to always be able to throw those kind of things.

krsriq replied:
Added it because it was requested (https://github.com/FakerPHP/Faker/pull/120#discussion_r546606173). You're right @GrahamCampbell - thanks!

localheinz commented:
@krsriq 

Running

```
$ composer-require-checker check
```

yields

```
ComposerRequireChecker 2.1.0@0c66698d487fcb5c66cf07108e2180c818fb2e72
The following unknown symbols were found:
+-----------------------------------------------------------+--------------------+
| unknown symbol                                            | guessed dependency |
+-----------------------------------------------------------+--------------------+
| Cake\ORM\TableRegistry                                    |                    |
| check_digit                                               |                    |
| ColumnMap                                                 |                    |
| CURLINFO_HTTP_CODE                                        | ext-curl           |
| CURLOPT_FILE                                              | ext-curl           |
| curl_close                                                | ext-curl           |
| curl_exec                                                 | ext-curl           |
| curl_getinfo                                              | ext-curl           |
| curl_init                                                 | ext-curl           |
| curl_setopt                                               | ext-curl           |
| Doctrine\Common\Persistence\Mapping\ClassMetadata         |                    |
| Doctrine\Common\Persistence\ObjectManager                 |                    |
| Doctrine\ODM\MongoDB\Mapping\ClassMetadata                |                    |
| Doctrine\ORM\Mapping\ClassMetadata                        |                    |
| DOMDocument                                               | ext-dom            |
| DOMElement                                                | ext-dom            |
| hash                                                      | ext-hash           |
| iconv                                                     | ext-iconv          |
| Mandango\Mandango                                         |                    |
| mb_strlen                                                 | ext-mbstring       |
| mb_strtolower                                             | ext-mbstring       |
| mb_strtoupper                                             | ext-mbstring       |
| mb_substr                                                 | ext-mbstring       |
| Propel                                                    |                    |
| PropelColumnTypes                                         |                    |
| Propel\Generator\Model\PropelTypes                        |                    |
| Propel\Runtime\Map\ColumnMap                              |                    |
| Propel\Runtime\Propel                                     |                    |
| Propel\Runtime\ServiceContainer\ServiceContainerInterface |                    |
| Spot\Locator                                              |                    |
| Spot\Mapper                                               |                    |
| Spot\Relation\BelongsTo                                   |                    |
| Transliterator                                            | ext-intl           |
+-----------------------------------------------------------+--------------------+
```

What do you think, should we suggest some of the extensions listed here as well?

pimjansen commented:
I think for now atleast suggest them which means we do not require yet. 

Then from 2.0 forward we can require and mark them as BC?

krsriq commented:
@localheinz - added `ext-curl` and `ext-mbstring` now to suggest.

pimjansen commented:
@localheinz i see codecov is blocking the merge due to untested lines. Will it only do this right now on this run since it is changed?

Since it is pretty lame to test if the extension is there or not

localheinz commented:
@pimjansen 

Taking a look!

localheinz reviewed (approved):
👍

localheinz commented:
Thank you, @krsriq!

krsriq commented:
Thank you!
